vyper/ast_utils.py

```diff
@@ -0,0 +1,141 @@
+import ast as python_ast
+
+import vyper.ast as vyper_ast
+from vyper.exceptions import (
+    ParserException,
+    SyntaxException,
+)
+from vyper.parser.parser_utils import (
+    annotate_and_optimize_ast,
+)
+from vyper.parser.pre_parser import (
+    pre_parse,
+)
+
+
+def parse_python_ast(source_code, node):
+    if isinstance(node, list):
+        o = []
+        for n in node:
+            o.append(
+                parse_python_ast(
+                    source_code=source_code,
+                    node=n,
+                )
+            )
+        return o
+    elif isinstance(node, python_ast.AST):
+        class_name = node.__class__.__name__
+        if hasattr(vyper_ast, class_name):
+            vyper_class = getattr(vyper_ast, class_name)
+            init_kwargs = {
+                'col_offset': getattr(node, 'col_offset', None),
+                'lineno': getattr(node, 'lineno', None),
+                'node_id': node.node_id,
+                'source_code': source_code
+            }
+            if isinstance(node, python_ast.ClassDef):
+                init_kwargs['class_type'] = node.class_type
```

Comment:
This is one place where the accumulation pattern using generators shines.  You can move this functionality into a small helper that looks like this.  Non mutative and easy to do conditional field logic.

```python
@to_dict
def _build_node_init_kwargs(node, source_code):
    yield (`col_offset': getattr(node, 'col_offset', None))
    yield ('lineno': getattr(node, 'lineno', None))
    yield ('node_id': node.node_id)
    yield ('source_code': source_code)
    if isinstance(node, python_ast.ClassDef):
        yield ('class_type', node.class_type)

    ... more for iterating over `node._fields`
```

Bonus points for using the `mypy_extensions.TypedDict` for this so you have strong type hinting.